ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/package.scala

```diff
@@ -30,9 +30,12 @@ import com.google.protobuf.ByteString
 package object kvutils {
 
   type Bytes = ByteString
-
   type DamlStateMap = Map[DamlStateKey, Option[DamlStateValue]]
 
+  type Fingerprint = Bytes
+  type DamlStateMapWithFingerprints = Map[DamlStateKey, (Option[DamlStateValue], Fingerprint)]
+  val FingerprintPlaceholder: Fingerprint = ByteString.EMPTY
```

Comment:
My understanding is that a fingerprint will always be present but we'll only use it in pre-execution; hence, I am defining using this value in the current (non-preexec) code path.

Reply:
We only need fingerprints for pre-execution, hence, let's have two separate `CommitContext` implementations -- one should have DAML state input with fingerprints whereas the other without fingerprints. We won't even need to be able to access read input keys for normal execution.

Reply:
It's a good idea, let's see if I can come up with something decent.

Reply:
I think the best moment to do that as cleanly as possible is after we bifurcate the code path into execution and pre-execution. Then, any left shared logic that still needs to discriminate between the two cases would take `Either[ExecutionCommitContext, PreexecutionCommitContext]` (both would inherit from a common trait with the common parts).

Reply:
Agreed regarding bifurcating the code paths for the two execution methods later. However, right now the simplest approach would be to have access to the same set of properties in the context and the driver (`KeyValueCommitting` or parent `Committer` class) can do whatever they want with those and also may do some pre-execution specific logic. I would not suggest introducing an `Either` for this as it's going to complicate code unnecessarily. Instead, we may want to define different set of steps for pre-execution vs. normal execution for the different committers and share the steps that are common.

Reply:
Yes, using separate steps can do it more cleanly. What's your suggestion for now though, leaving as it is if I understand correctly?

Reply:
Yes, let's leave it as-is for now and do a refactoring at a later step.

Reply:
@miklos-da If everything else looks at this stage, could you approve it?